cpp/include/cudf/interop.hpp

```diff
@@ -130,6 +140,298 @@ using unique_schema_t = std::unique_ptr<ArrowSchema, void (*)(ArrowSchema*)>;
+ *
+ */
+using unique_table_view_t =
+  std::unique_ptr<cudf::table_view, custom_view_deleter<cudf::table_view>>;
+
+/**
+ * @brief typedef for a unique_ptr to a `cudf::column_view` with custom deleter
+ *
+ */
+using unique_column_view_t =
+  std::unique_ptr<cudf::column_view, custom_view_deleter<cudf::column_view>>;
+
+/**
+ * @brief A wrapper around ArrowDeviceArray data used for flexible lifetime management.
+ */
+struct arrow_array_container;
+
+/**
+ * @brief Helper function to generate empty column metadata (column with no
+ * name) for arrow conversion.
+ *
+ * This function is helpful for internal conversions between host and device
+ * data using existing arrow functions. It is also convenient for external
+ * usage of the libcudf Arrow APIs to produce the canonical mapping from cudf
+ * column names to Arrow column names (i.e. empty names with appropriate
+ * nesting).
+ *
+ * @param input The column to generate metadata for
+ * @return The metadata for the column
+ */
+cudf::column_metadata get_column_metadata(cudf::column_view const& input);
+
+/**
+ * @brief Helper function to generate empty table metadata (all columns with no
+ * names) for arrow conversion.
+ *
+ * This function is helpful for internal conversions between host and device
+ * data using existing arrow functions. It is also convenient for external
+ * usage of the libcudf Arrow APIs to produce the canonical mapping from cudf
+ * column names to Arrow column names (i.e. empty names with appropriate
+ * nesting).
+ *
+ * @param input The table to generate metadata for
+ * @return The metadata for the table
+ */
+std::vector<cudf::column_metadata> get_table_metadata(cudf::table_view const& input);
+
+/**
+ * @brief A standard interchange medium for ArrowDeviceArray data in cudf.
+ *
+ * This class provides a way to work with ArrowDeviceArray data in cudf without
+ * sacrificing the APIs expected of a cudf column. On the other end, it
+ * provides the shared lifetime management expected by arrow consumers rather
+ * than the single-owner mechanism of cudf::column.
+ */
+class arrow_column {
+ public:
+  /**
+   * @brief Construct a new arrow column object
+   *
+   * The input array will be moved into the arrow_column, so it is no longer
+   * suitable for use afterwards.
+   *
+   * @param schema Arrow schema for the column
+   * @param input ArrowDeviceArray data for the column
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used for any allocations during conversion
+   */
+  arrow_column(ArrowSchema const* schema,
+               ArrowDeviceArray* input,
+               rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+               rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref());
+
+  /**
+   * @brief Construct a new arrow column object
+   *
+   * The input array will be released, so it is no longer suitable for use
+   * afterwards.
```

Comment:
since this requires a copy from host -> device, are we just saying we'll release it for consistency even though it's not technically needed?

Reply:
Yes, exactly. I decided consistent semantics across the board were probably preferable. With the current functions I find myself looking at the documentation consistently to figure out how it behaves. If you feel strongly about not doing this in the CPU case I could certainly be convinced not to release here, though.

Reply:
I'm fine with the consistency, just wanted to confirm that was the reason.